packages/ibm-products/src/components/Datagrid/Datagrid/addons/Filtering/FilterProvider.js

```diff
@@ -27,22 +34,100 @@ const EventEmitter = {
   },
 };
 
+const removeFilterItem = (state, index) => state.splice(index, 1);
+
+const updateFilterState = (state, type, value) => {
+  if (type === CHECKBOX) {
+    return;
+  }
+  if (type === DATE) {
+    const filterTagIndex = state.findIndex(
+      (val) =>
+        formatDateRange(val.value[0], val.value[1]) ===
+        formatDateRange(value[0], value[1])
+    );
+    return removeFilterItem(state, filterTagIndex);
+  }
+  const filterTagIndex = state.findIndex((val) => val.value === value);
+  return removeFilterItem(state, filterTagIndex);
+};
+
+export const clearSingleFilter = ({ key, value }, setAllFilters, state) => {
+  const tempState = [...state.filters];
+  tempState.forEach((f, filterIndex) => {
+    if (f.id === key) {
+      const filterValues = f.value;
+      const filterType = f.type;
+      updateFilterState(tempState, filterType, value);
+      if (filterType === CHECKBOX) {
+        /**
+          When all checkboxes of a group are all unselected the value still exists in the filtersObjectArray
+          This checks if all the checkboxes are selected = false and removes it from the array
+        */
+        const valueIndex = filterValues.findIndex((val) => val.id === value);
+        filterValues[valueIndex].selected = false;
+        const updatedFilterObject = {
+          ...f,
+          value: [...filterValues],
+        };
+        tempState[filterIndex] = updatedFilterObject;
+        const index = tempState.findIndex((filter) => filter.id === key);
+
+        // If all the selected state is false remove from array
+        const shouldRemoveFromArray = tempState[index].value.every(
+          (val) => val.selected === false
+        );
+
+        if (shouldRemoveFromArray) {
+          removeFilterItem(tempState, index);
+        }
+      }
+    }
+  });
+  setAllFilters(tempState);
+};
+
+const handleSingleFilterRemoval = (key, value) => {
+  EventEmitter.dispatch(CLEAR_SINGLE_FILTER, { key, value });
+};
+
+const formatDateRange = (startDate, endDate) => {
+  const startDateObj = new Date(startDate);
+  const endDateObj = new Date(endDate);
+  return `${startDateObj.toLocaleDateString()} - ${endDateObj.toLocaleDateString()}`;
+};
+
 const prepareFiltersForTags = (filters) => {
   const tags = [];
 
   filters.forEach(({ id, type, value }) => {
+    const sharedFilterProps = {
+      filter: true,
+      onClose: () => handleSingleFilterRemoval(id, value),
+    };
+
     if (type === DROPDOWN || type === RADIO || type === NUMBER) {
-      tags.push({ key: id, value });
+      tags.push({
+        key: id,
+        value,
+        ...sharedFilterProps,
+      });
     } else if (type === DATE) {
       const [startDate, endDate] = value;
       tags.push({
         key: id,
-        value: `${startDate.toLocaleDateString()} - ${endDate.toLocaleDateString()}`,
+        value: formatDateRange(startDate, endDate),
```

Comment:
I wonder if this is a function that we can also let the dev's decide how to format it and have this as a backup. WDYT? 🤔

Reply:
Yes, definitely agree. Similar to `renderLabel` for the `FilterSummary`.

Reply:
Added [here](https://github.com/carbon-design-system/ibm-products/pull/3582/commits/cd8c9ab5e4dc8fb1c7fa36a78e14aa4cec8d147d)